# Next cell is for running on GPU
開啟檔案後預設的運行單元為CPU，若要改成GPU則可以按下面兩圖操作：https://mattwang44.github.io/en/images/DL/colab/5.png,
https://mattwang44.github.io/en/images/DL/colab/6.png

更改後可以用運行這一段程式碼確認GPU運行正常

In [2]:
import tensorflow as tf
device_name = tf.test.gpu_device_name()
if device_name != '/device:GPU:0':
  raise SystemError('GPU device not found')
print('Found GPU at: {}'.format(device_name))

Found GPU at: /device:GPU:0


# Next cell is for using TPU.
See: https://colab.research.google.com/notebooks/tpu.ipynb#scrollTo=G2pjP0c5VqhK

In [0]:
import os
import pprint
%tensorflow_version 1.x
import tensorflow as tf

if 'COLAB_TPU_ADDR' not in os.environ:
  print('ERROR: Not connected to a TPU runtime; please see the first cell in this notebook for instructions!')
else:
  tpu_address = 'grpc://' + os.environ['COLAB_TPU_ADDR']
  print ('TPU address is', tpu_address)

  with tf.Session(tpu_address) as session:
    devices = session.list_devices()
    
  print('TPU devices:')
  pprint.pprint(devices)

ERROR: Not connected to a TPU runtime; please see the first cell in this notebook for instructions!


運行下面的程式碼，會出現鏈接，點開鏈接，選擇自己的谷歌帳號，將得到的代碼輸入鏈接下面的框中即可。

In [7]:
!apt-get install -y -qq software-properties-common python-software-properties module-init-tools
!add-apt-repository -y ppa:alessandro-strada/ppa 2>&1 > /dev/null
!apt-get update -qq 2>&1 > /dev/null
!apt-get -y install -qq google-drive-ocamlfuse fuse
from google.colab import auth
auth.authenticate_user()
from oauth2client.client import GoogleCredentials
creds = GoogleCredentials.get_application_default()
import getpass
!google-drive-ocamlfuse -headless -id={creds.client_id} -secret={creds.client_secret} < /dev/null 2>&1 | grep URL
vcode = getpass.getpass()
!echo {vcode} | google-drive-ocamlfuse -headless -id={creds.client_id} -secret={creds.client_secret}

E: Package 'python-software-properties' has no installation candidate
Selecting previously unselected package google-drive-ocamlfuse.
(Reading database ... 145605 files and directories currently installed.)
Preparing to unpack .../google-drive-ocamlfuse_0.7.14-0ubuntu1~ubuntu18.04.1_amd64.deb ...
Unpacking google-drive-ocamlfuse (0.7.14-0ubuntu1~ubuntu18.04.1) ...
Setting up google-drive-ocamlfuse (0.7.14-0ubuntu1~ubuntu18.04.1) ...
Processing triggers for man-db (2.8.3-2ubuntu0.1) ...
Please, open the following URL in a web browser: https://accounts.google.com/o/oauth2/auth?client_id=32555940559.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive&response_type=code&access_type=offline&approval_prompt=force
··········
Please, open the following URL in a web browser: https://accounts.google.com/o/oauth2/auth?client_id=32555940559.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope

In [3]:
!nvidia-smi # this will work if run on GPU, but fail if run on TPU

Wed Nov 27 16:54:50 2019       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.33.01    Driver Version: 418.67       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   34C    P0    32W / 250W |    265MiB / 16280MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU  

掛載谷歌雲端硬碟：以下程式碼則可以幫我們把虛擬機和自己的Google Drive連接，執行後虛擬機目錄下會多出一個名為Drive(可自行更改)*的資料夾，內容即為自己的雲端資料。*

In [0]:
!mkdir -p Drive 
!google-drive-ocamlfuse Drive -o nonempty

每次修改了雲盤中的文件以後都最好運行一次，運行了沒有任何提示便成功了。

若是出現Warning: apt-key output should not be parsed (stdout is not a terminal)警告標語，代表和自己雲端連接的權限已經到期，重新執行上面這段code就可以重新連接。

設置工作路徑：

In [0]:
import os 
os.chdir( 'Drive' )

In [6]:
!ls 指導的學生/2019_吳玟萱/2019_1104_Joke_Datasets

 cache_dir
 Excel2txt.py
 interpreting-text-classification-models-with-eli5.ipynb
 JokeHumorLevel.ipynb
 Joke_Length.ipynb
 JokeMotivation.ipynb
 JokeMotivation_random.ipynb
 JokeSkill.ipynb
 JokeSkill_random.ipynb
 mlabel_corpora
 Multi-Label-Text-Classification-master
 out
 __pycache__
 runs
 Set_Random_Train_Motivation.ipynb
 Set_Random_Train_Skill.ipynb
 Shuffle_Matrix.pl
 SimpleTransformer_HumorLevel.ipynb
 SimpleTransformer_Motivation_colab.ipynb
 SimpleTransformer_Motivation_random_colab.ipynb
 SimpleTransformer_Skill_colab.ipynb
 SimpleTransformer_Skill_random_colab.ipynb
 simpletransformers-master
 SNsanafonGyou.ttf
 Stopwords.py
 刪除笑話統計表.xlsx
'~$笑話標記.xlsx'
 笑話標記.xlsx
 笑話資料庫.xlsx
 負範例.xlsx


In [0]:
!pip install pytorch cudatoolkit=10.0 -c pytorch

ERROR: Could not open requirements file: [Errno 2] No such file or directory: 'pytorch'


In [0]:
!pip show pytorch

In [0]:
!pip --version

pip 19.3.1 from /usr/local/lib/python3.6/dist-packages/pip (python 3.6)


In [7]:
!pip install transformers

In [8]:
!pip install seqeval

In [9]:
!pip install tensorboardx

In [10]:
!pip install simpletransformers

In [0]:
from simpletransformers.classification import MultiLabelClassificationModel
import pandas as pd
import sys, time

In [0]:
data_path = "指導的學生/2019_吳玟萱/2019_1104_Joke_Datasets/mlabel_corpora/JokeMotivation.txt"
train_random_path = "指導的學生/2019_吳玟萱/2019_1104_Joke_Datasets/mlabel_corpora/JokeMotivation_train_random.txt"

In [13]:
# set global variables: df
df = pd.read_csv(data_path, delimiter="\t")
#data_raw = df.loc[np.random.choice(data_raw.index, size=2000)]
train_random = pd.read_csv(train_random_path, delimiter="\t")
print(df.shape, train_random.shape) # same as data_raw.shape in Jupyter

(3365, 9) (1691, 9)


In [14]:
from sklearn.model_selection import train_test_split
# ID=L1850 為分界，之前：吳玟萱，之後：黃亭筠，均為中文系同一屆
train, test = train_test_split(df, train_size=1691, shuffle=False) 
# (tempararily) set global variables: train, test 
print(train_random.shape)
print(train.shape, test.shape)
print(train.columns)
print(train_random.columns)

(1691, 9)
(1691, 9) (1674, 9)
Index(['ID', 'Title', 'Content', 'Affinity', 'Self_Improvement', 'Attack',
       'Self_Depression', 'Taboo', 'Others'],
      dtype='object')
Index(['ID', 'Title', 'Content', 'Affinity', 'Self_Improvement', 'Attack',
       'Self_Depression', 'Taboo', 'Others'],
      dtype='object')


## set global variables: train_df, test_df

In [17]:
# https://stackoverflow.com/questions/34682828/extracting-specific-selected-columns-to-new-dataframe-as-a-copy
# selected column index: 1, 6, 7
# new = old.iloc[: , [1, 6, 7]].copy() 
# new = old[['A', 'C', 'D']].copy()

train_df = train_random[['Content']].copy()
#train_df = train_df.drop(labels = ['ID', 'Title', 'Content'], axis=1)
#https://stackoverflow.com/questions/33098383/merge-multiple-column-values-into-one-column-in-python-pandas
train_df['labels'] = train_random[train_random.columns[3:]].apply(
    lambda x: list(x), axis=1)
#    lambda x: ','.join(x.dropna().astype(str)), axis=1)
#train_df['labels'] = train_df[['labels']].apply(lambda x: x.split(','), axis=1)
# train.columns = ['ID', 'Title', 'Content', 'Pun', 'Exaggeration', 'Anthropomorphism', 'Bridge_Inference', 'Illogical', 'Irony', 'Imitation', 'Others']
#train_df = train_df.drop(labels = train.columns, axis=1)


test_df = test.loc[:, ['Content']].copy()
#https://stackoverflow.com/questions/33098383/merge-multiple-column-values-into-one-column-in-python-pandas
test_df['labels'] = test.loc[:, test.columns[3:]].apply(
    lambda x: list(x), axis=1)
#    lambda x: ','.join(x.dropna().astype(str)), axis=1)
#test_df['labels'] = test_df[['labels']].apply(lambda x: x.split(','), axis=1)
#test_df = test_df.drop(labels = test.columns, axis=1)

print(f'{train_random.shape}, {test.shape}, {train_df.shape}, {test_df.shape}')
print()
print(train_df.head())
print(test_df.head())

(1691, 9), (1674, 9), (1691, 2), (1674, 2)

                                             Content              labels
0  員工：老闆，您必須幫我加薪，已經有三家公司在找我了！     老闆：哪三家？     員工：...  [0, 0, 0, 0, 1, 0]
1  某市政府辦公大樓落成，門口缺副對聯。     副市長揮毫     上聯：說實話辦實事一身正氣...  [0, 0, 0, 0, 0, 1]
2  中午老闆視察自己的建築工地時，發現有個人在角落玩手機。     老闆：你月薪多少？     ...  [0, 0, 0, 0, 0, 1]
3  一天，一位法官的女友看見兩個蚊子，便叫法官打死。     只見法官只把那個肚子飽飽的蚊子打死...  [0, 0, 1, 0, 0, 0]
4  辦公室中兩位女同事吵起來了。     經理忍無可忍：「太不像話了！現在是什麼情況？你們把原因...  [0, 0, 0, 0, 0, 1]
                       Content              labels
1691  即使你已名花有主，我也要移花接木把你移到我身邊。  [0, 0, 0, 0, 0, 1]
1692              錢不是問題，問題是沒錢！  [0, 0, 0, 1, 0, 0]
1693    自從我變成了狗屎，就再也沒有人踩在我頭上了。  [0, 0, 0, 1, 0, 0]
1694         蘋果最光輝的一刻就是砸在牛頓頭上！  [0, 0, 0, 0, 0, 1]
1695   其實上課睡覺不是我們的錯，只怪地球吸引力太大。  [0, 0, 0, 0, 0, 1]


In [0]:
# Train and Evaluation data needs to be in a Pandas Dataframe containing at least two columns, a 'text' and a 'labels' column. The `labels` column should contain multi-hot encoded lists.

#train_data = [['Example sentence 1 for multilabel classification.', [1, 1, 1, 1, 0, 1]]] + [['This is another example sentence. ', [0, 1, 1, 0, 0, 0]]]
#train_df = pd.DataFrame(train_data, columns=['text', 'labels'])
#train_df = pd.DataFrame(train_data)

#eval_data = [['Example eval sentence for multilabel classification.', [1, 1, 1, 1, 0, 1]], ['Example eval senntence belonging to class 2', [0, 1, 1, 0, 0, 0]]]
#eval_df = pd.DataFrame(eval_data)

#print(train_df.head())
#print(eval_df.head())


In [0]:
# Create a MultiLabelClassificationModel
model = MultiLabelClassificationModel('bert', 'bert-base-chinese', 
#model = MultiLabelClassificationModel('bert', '/Users/sam/GoogleDrive/BERT/bert/chinese_L-12_H-768_A-12/', 
    num_labels=6, use_cuda=True, 
    args={'reprocess_input_data':True, 'overwrite_output_dir':True, 'num_train_epochs':10,
          'max_seq_length': 512, 'train_batch_size': 8, 'eval_batch_size': 8,
          'output_dir':'outputs/Motivation_random'})
# You can set class weights by using the optional weight argument


In [19]:
# https://ipywidgets.readthedocs.io/en/stable/user_install.html
# conda update -n base -c defaults conda # this will update conda version for all virtualenv
# conda install -c conda-forge ipywidgets

stime = time.time()
# Train the model
model.train_model(train_df)
print('training time: %1.2f'%(time.time()-stime))


Converting to features started.


Running loss: 0.390003

Running loss: 0.454937

Running loss: 0.395360

Running loss: 0.312726

Running loss: 0.097068

Running loss: 0.291703

Running loss: 0.322828

Running loss: 0.116384

Running loss: 0.070736

Running loss: 0.019452
Training of bert model complete. Saved to outputs/Motivation_random.
training time: 1132.42


In [20]:
stime = time.time()
# Evaluate the model
result, model_outputs, wrong_predictions = model.eval_model(test_df)
print(result)
print(model_outputs) # this is the raw prediction
print('Wrong Prediction:')
print(wrong_predictions)
print('Eval time: %1.2f'%(time.time()-stime))

Converting to features started.



{'LRAP': 0.6632666268418966, 'eval_loss': 0.5011732717887276}
[[0.01638897 0.00920349 0.00700845 0.01603195 0.01569277 0.9890082 ]
 [0.14329435 0.07624249 0.03693737 0.02707085 0.15651155 0.04465185]
 [0.52249855 0.09687398 0.01772476 0.16675964 0.10126244 0.02259376]
 ...
 [0.00901666 0.00425899 0.00468921 0.00663711 0.03322103 0.98655117]
 [0.01464027 0.00501558 0.00580285 0.00848642 0.01056523 0.9792037 ]
 [0.0406676  0.01186764 0.00316319 0.04875422 0.14957969 0.88764817]]
Wrong Prediction:
[<simpletransformers.classification.classification_utils.InputExample object at 0x7f913c7c4390>, <simpletransformers.classification.classification_utils.InputExample object at 0x7f913c7c44a8>, <simpletransformers.classification.classification_utils.InputExample object at 0x7f913c7c4208>, <simpletransformers.classification.classification_utils.InputExample object at 0x7f913c7c4630>, <simpletransformers.classification.classification_utils.InputExample object at 0x7f903afafda0>, <simpletransformer

In [0]:
from sklearn.metrics import accuracy_score, roc_auc_score, roc_curve
from sklearn.metrics import precision_recall_fscore_support
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
import numpy as np 

In [0]:
# define the evaluation metrics
def tcfunc(x, n=4): # trancate a number to have n decimal digits
    d = '0' * n
    d = int('1' + d)
# https://stackoverflow.com/questions/4541155/check-if-a-number-is-int-or-float
    if isinstance(x, (int, float)): return int(x * d) / d
    return x

def print_cls_report(y_true, prediction):
    print('Test accuracy is %1.4f'%(accuracy_score(y_true, prediction)))

    print(classification_report(y_true, prediction))
    
    # http://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_fscore_support.html
    print("\tPrecision\tRecall\tF1\tSupport")
    (Precision, Recall, F1, Support) = list(map(tcfunc, 
        precision_recall_fscore_support(y_true, prediction, average='micro')))
    print("Micro\t{}\t{}\t{}\t{}".format(Precision, Recall, F1, Support))

    (Precision, Recall, F1, Support) = list(map(tcfunc, 
        precision_recall_fscore_support(y_true, prediction, average='macro')))
    print("Macro\t{}\t{}\t{}\t{}".format(Precision, Recall, F1, Support))
    
#    if True:
    if False:
        print(confusion_matrix(y_true, prediction))
        try: 
            print(classification_report(y_true, prediction, digits=4))
        except ValueError:
            print('May be some category has no predicted samples')
        show_confusion_matrix(prediction)


    print(f'y_true.shape={y_true.shape}, prediction.shape={prediction.shape}')
    #print(y_true.head())
    #print(prediction[0:6])

    # https://stackoverflow.com/questions/152580/whats-the-canonical-way-to-check-for-type-in-python
    pred = prediction
    if not isinstance(pred, np.ndarray): pred = prediction.toarray()
    print(type(y_true), type(prediction), type(pred))
    try:
        # https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html
        print('macro roc_auc_score is %1.4f'%(roc_auc_score(y_true, pred, average='macro'))) # default average=’macro’
        print('micro roc_auc_score is %1.4f'%(roc_auc_score(y_true, pred, average='micro')))
    except:
        print("roc_auc_score error!!!")
    try:
        fpr, tpr, thresholds = roc_curve(y_true, pred)
        print(f'fpr={fpr}\ntpr={tpr}\nthresholds={thresholds}')
    except:
        print('roc_curve error!!!')

In [23]:
type(model_outputs)
# https://stackoverflow.com/questions/35215161/most-efficient-way-to-map-function-over-numpy-array

# Obtain array of square of each element in x
threshold = lambda x: 1 if x>0.5 else 0 
predictions = np.array([[threshold(x) for x in row] for row in model_outputs])
print(type(predictions), type(test_df['labels']))
print('predictions:')
print(predictions)
y_true = np.array([[x for x in row] for row in test_df['labels'] ])
print(f'y_true, {type(y_true)}')
print(y_true)

<class 'numpy.ndarray'> <class 'pandas.core.series.Series'>
predictions:
[[0 0 0 0 0 1]
 [0 0 0 0 0 0]
 [1 0 0 0 0 0]
 ...
 [0 0 0 0 0 1]
 [0 0 0 0 0 1]
 [0 0 0 0 0 1]]
y_true, <class 'numpy.ndarray'>
[[0 0 0 0 0 1]
 [0 0 0 1 0 0]
 [0 0 0 1 0 0]
 ...
 [0 0 0 0 0 1]
 [0 0 0 0 0 1]
 [0 0 0 0 0 1]]


In [24]:
print_cls_report(y_true, predictions)

Test accuracy is 0.4630
              precision    recall  f1-score   support

           0       0.00      0.00      0.00        19
           1       0.00      0.00      0.00        16
           2       0.04      0.01      0.02       143
           3       0.00      0.00      0.00        17
           4       0.11      0.13      0.12       229
           5       0.74      0.61      0.67      1254

   micro avg       0.57      0.47      0.52      1678
   macro avg       0.15      0.12      0.13      1678
weighted avg       0.57      0.47      0.52      1678
 samples avg       0.47      0.47      0.47      1678

	Precision	Recall	F1	Support
Micro	0.5715	0.4737	0.518	None
Macro	0.148	0.1249	0.1342	None
y_true.shape=(1674, 6), prediction.shape=(1674, 6)
<class 'numpy.ndarray'> <class 'numpy.ndarray'> <class 'numpy.ndarray'>
macro roc_auc_score is 0.4900
micro roc_auc_score is 0.7013
roc_curve error!!!


/usr/local/lib/python3.6/dist-packages/sklearn/metrics/classification.py:1437: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/classification.py:1437: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in samples with no predicted labels.
  'precision', 'predicted', average, warn_for)
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/classification.py:1439: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in samples with no true labels.
  'recall', 'true', average, warn_for)


# When epochs=10
###batch_size=8, length=512

    1. Running loss: 0.373369
    2. Running loss: 0.316603
    3. Running loss: 0.211718
    4. Running loss: 0.236091
    5. Running loss: 0.044653
    6. Running loss: 0.113206
    7. Running loss: 0.088623
    8. Running loss: 0.018922
    9. Running loss: 0.065161
    10. Running loss: 0.054991

    Training of bert model complete. 
    Saved to outputs/.

training time: 1142.57 (training time: 2124.60, where batch_size=4)

Test accuracy is 0.2168
            
              precision    recall  f1-score   support
           0       0.32      0.56      0.41       261
           1       0.00      0.00      0.00        60
           2       0.47      0.53      0.49        40
           3       0.19      0.25      0.21       291
           4       0.34      0.43      0.38       486
           5       0.00      0.00      0.00        35
           6       0.13      0.07      0.09       110
           7       0.55      0.04      0.07       478

    micro avg       0.30      0.27      0.28      1761
    macro avg       0.25      0.23      0.21      1761
    weighted avg       0.34      0.27      0.24      1761
    samples avg       0.26      0.27      0.26      1761

	Precision	Recall	F1	Support
    Micro	0.2969	0.2703	0.2829	None
    Macro	0.2485	0.2345	0.2075	None

y_true.shape=(1674, 8), prediction.shape=(1674, 8)
  
    macro roc_auc_score is 0.5640
    micro roc_auc_score is 0.5867
  roc_curve error!!!

### when batch_size=4
Test accuracy is 0.2139 
              
              precision    recall  f1-score   support
           0       0.37      0.46      0.41       261
           1       0.00      0.00      0.00        60
           2       0.39      0.55      0.45        40
           3       0.20      0.48      0.28       291
           4       0.33      0.39      0.36       486
           5       0.00      0.00      0.00        35
           6       0.14      0.08      0.10       110
           7       0.46      0.05      0.09       478

    micro avg       0.28      0.29      0.28      1761
    macro avg       0.23      0.25      0.21      1761
    weighted avg       0.32      0.29      0.25      1761
    samples avg       0.26      0.29      0.27      1761

	Precision	Recall	F1	Support
    Micro	0.2806	0.2856	0.2831	None
    Macro	0.2347	0.2511	0.211	None
  
    y_true.shape=(1674, 8), prediction.shape=(1674, 8)
    macro roc_auc_score is 0.5651
    micro roc_auc_score is 0.5874
    roc_curve error!!!


# When epochs=4
training time: 844.57

Test accuracy is 0.2073

              precision    recall  f1-score   support

           0       0.37      0.46      0.41       261
           1       0.00      0.00      0.00        60
           2       0.43      0.50      0.46        40
           3       0.21      0.07      0.10       291
           4       0.34      0.46      0.39       486
           5       0.00      0.00      0.00        35
           6       0.17      0.01      0.02       110
           7       0.43      0.02      0.04       478


    micro avg       0.34      0.22      0.27      1761
    macro avg       0.24      0.19      0.18      1761
    weighted avg       0.32      0.22      0.21      1761
    samples avg       0.23      0.22      0.23      1761

	Precision	Recall	F1	Support
    Micro	0.3441	0.2231	0.2707	None
    Macro	0.2428	0.189	0.1768	None
    y_true.shape=(1674, 8), prediction.shape=(1674, 8)
    macro roc_auc_score is 0.5578
    micro roc_auc_score is 0.5794
    roc_curve error!!!

In [0]:
stime = time.time()
predictions, raw_outputs = model.predict(['即使你已名花有主，我也要移花接木把你移到我身邊。', '錢不是問題，問題是沒錢！'])
print(predictions)
print(raw_outputs)
print('Predict time: %1.2f'%(time.time()-stime))

Converting to features started.



[[0, 0, 0, 0, 0, 0, 0, 1], [0, 0, 0, 0, 0, 0, 0, 1]]
[[0.06845585 0.06373166 0.02769983 0.3163353  0.04200668 0.06746653
  0.08277455 0.5324532 ]
 [0.05592972 0.06493436 0.0297754  0.20465398 0.0532307  0.04532032
  0.07499086 0.5724112 ]]
Predict time: 1.82
